### PR TITLE
In the X.509 SVID draft spec, require strong signature algorithms.

### PR DESCRIPTION
**Pull request by briansmith**

### CODE REVIEW

**Comment by briansmith** [comment]
This is a proposal. Note that we might benefit from having different requirements for issuers vs. verifiers. For example, we may want to prohibit issuers from using Ed25519 until it is more widely deployed, but require verifiers to support it ASAP.

**Comment by rra** [comment]
I'm a little dubious about coding strength requirements into this spec because I'm not sure it will successfully track the evolution of SSL best practices.  Most of the things we're saying here are unlikely to change very quickly, but this will definitely change.  So encoding this sort of thing in the standard is signing up for an obligation to continually update these requirements.

Maybe it makes sense to do that, but I'd personally lean towards deferring to general TLS best practices rather than trying to copy them here, since there's nothing specific about SPIFFE in this.  The same guidelines could apply to any user of TLS for any purpose.

**Comment by briansmith** [comment]
@rra That's a good point. At the same time, it's important that implementations know which algorithms, which curves, and which keys sizes they are required to support for interop.

Since there are no legacy SPIFFE certificates, there's no need to support legacy algorithms. Yet, the default setting for many OpenSSL CAs will result in legacy algorithms and key sizes being used, e.g. RSA 1024 using PKCS#1 with SHA-1.

Note that these guidelines are borrowed from the guidelines for Mozilla's CA program at https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/policy/#algorithms and other guidelines at IETF, all of which I helped develop. This also happens to be exactly the set of algorithms my own implementation supports.

**Comment by rra** [comment]
Right, I have no objections to the guidelines as stated.  They seem quite reasonable.  And completely agreed that SPIFFE probably doesn't have a legacy issue.  (I could construct some scenarios in which this is not strictly true with dual-use certificates, but it's an unlikely edge case.)

My only concern here is about long-term maintainability, since I've seen recommendations like this stay static in other standards and then fall embarassingly out of date.  But presumably the hope is that this standard will be more dynamic, so if the core developers are comfortable with committing to that ongoing maintenance, there's value in saying this.

It does make implementations more complex because they have to check a bunch of stuff to be conforming, but the security is probably worth it.

File: standards/X509-SVID.md [hunk 1]
```diff
@@ -90,14 +90,23 @@ The validation of trust in a given SVID is based on standard X.509 path validati
...
 
+When verifying signatures (e.g. when verifying the signature of a certificate using its issuer's public key, and when verifying a signature in a TLS handshake using the leaf certificate's public key) the validator MUST ensure that the signature uses a secure signature algorithm. In particular, all signatures MUST use one of the following signature algorithms:
+
+* ECDSA using the P-256 curve and using SHA-256 as the digest algorithm.
+* ECDSA using the P-384 curve and using SHA-384 as the digest algorithm.
+* RSA using PSS (not PKCS#1), with an RSA public key having a modulus of at least 2047 bits and at most 8192 bits, and an odd public exponent of at least 65537, and using one of SHA-256, SHA-384, or SHA-512 as the digest algorithm.
+* Ed25519 (using SHA-512) as specified in https://tools.ietf.org/html/draft-ietf-curdle-pkix-05.
```
Comment (azdagron) [inline standards/X509-SVID.md:106]: This should probably be updated (after merge is ok) to the official RFC link: https://tools.ietf.org/html/rfc8410

**Comment by evan2645** [comment]
We had a lengthy conversation about this in the last SIG-Spec call (meeting notes [here](https://docs.google.com/document/d/1f64vbyn5sOb8Mr1H3mGGGul3vTKo4r6cTBcUV3N9OFo/edit?pli=1)) as part of an effort to clean out stale prs/issues.

The consensus there echoes the concern already stated here, relating to the ongoing maintenance and hygiene of the list of supported signature algorithms. Even beyond ensuring that they are kept up to date, concern was raised over what happens interop-wise when we do need to update this list. Finally, it was acknowledged that there is unlikely a one-size-fits-all solution as governmental and regulatory bodies may impose mandates on key types and signature algorithms.

While the concept of requiring strong signature algorithms is one that is well-liked, the practical implications are large enough that the preference of the group is to omit the requirement from the X509-SVID spec altogether. We may revisit this at a later date in the form of a versioned "SPIFFE mTLS Profile" which can not only bolster security as this proposed change would, but also have a positive impact on interoperability as it relates to X509-SVID's primary use case.

**Comment by briansmith** [comment]
The consequence of this is that two fully-conformant SPIFFE implementations won't necessarily interoperate, and won't necessarily be secure. IMO the point of a spec is to ensure interop and correctness.

**Comment by evan2645** [comment]
> The consequence of this is that two fully-conformant SPIFFE implementations won't necessarily interoperate, and won't necessarily be secure. IMO the point of a spec is to ensure interop and correctness.

Yes, this point is understood. We've discussed it several times as part of our weekly calls, and are still struggling to find the right balance between interop and flexibility. By excluding this requirement, we may find that interop problems are encountered, though the likelihood and magnitude of those problems are hard for us to reason about. Contrarily, we may find that by including this requirement, we damage adoption due to the difficulties related to conforming.

Certainly we're very interested in having our cake and eating it too. Members of the SPIFFE SIG-Spec group have expressed interest in discussing this further with you, particularly with regards to your past experiences in this area. Any chance you'd be open to [joining sometime soon](https://github.com/spiffe/spiffe/blob/master/community/sig-spec/README.md) to help us further understand our options and their implications?

**Comment by briansmith** [comment]
> Any chance you'd be open to [joining sometime soon](https://github.com/spiffe/spiffe/blob/master/community/sig-spec/README.md) to help us further understand our options and their implications?

Currently my work leaves me quite far removed from this, but thanks for the offer.

**Pull request closed without merging**